### PR TITLE
[MNT] Add pre-commit, drop python 3.8, add python 3.12, add yamllint, add toml-sort

### PR DESCRIPTION
**Pull request by mscheltienne**
`pre-commit` is optional. 
If you want to use it:

```
pip install pre-commit
cd pycrostates  # navigate to git clone
pre-commit install
```

You can run on all files with `pre-commit run --all-files`.

If you want to remove it:

```
cd pycrostates  # navigate to git clone
pre-commit uninstall
```

### CODE REVIEW

**Comment by mscheltienne** [comment]
@vferat Can you go on https://pre-commit.ci/ and enable this CI for this repository?
It will auto-fix style with `black`, `isort`, `ruff` and `bibclean` in PRs.

**Comment by mscheltienne** [comment]
Ok, that should do it. I also fixed some MNE 1.6 compatibility issues (again), and I added a linter for `pyproject.toml` and for the YAML files.

**Comment by mscheltienne** [comment]
@vferat This time it should be green. In the end this PR does:

- drop python 3.8 and add python 3.12 to CIs / supported version
- add yamllint to lint YAML files
- add toml-sort to sort `pyproject.toml`
- add `pre-commit` hooks and configuration for https://pre-commit.ci/ (you'll need to enable that one on the CI website, and then a bot will automatically fix bad style in PRs)
- bump the minimum MNE version to 1.2 and fix some `check_version`

And then.. I decided to turn warnings into errors in `pytest`. So a bunch of fixes + a better configuration of `pytest`, especially for `matplotlib`. 

---------------

I added a GH workflows to help me debug the Azure workflows. The output is cleaner, the configuration file shorter, and the macOS runner faster. I'm inclined to drop the AZP workflows entirely in favor of GH, as MNE did. WDYT?
Before merging this PR, one of the 2 CI provided should be removed/disabled.

File: pycrostates/conftest.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import pytest
+from mne import set_log_level as set_log_level_mne
+
+from pycrostates import set_log_level
+
+
+def pytest_configure(config):
+    """Configure pytest options."""
+    for fixture in ("matplotlib_config", "close_all"):
+        config.addinivalue_line("usefixtures", fixture)
```
Comment (mscheltienne) [inline pycrostates/conftest.py:13]:
```suggestion
    config.addinivalue_line("usefixtures", "matplotlib_config")
```

File: pycrostates/conftest.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import os
+import warnings
+
+import pytest
+from mne import set_log_level as set_log_level_mne
+
+from pycrostates import set_log_level
```
Comment (mscheltienne) [inline pycrostates/conftest.py:7]:
```suggestion
from pycrostates import logger, set_log_level
```

File: pycrostates/conftest.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    ignore:Use setlocale.*instead:DeprecationWarning
+    ignore:datetime\.datetime\.utcnow.*is deprecated.*:DeprecationWarning
+    """
+    for warning_line in warnings_lines.split("\n"):
+        warning_line = warning_line.strip()
+        if warning_line and not warning_line.startswith("#"):
+            config.addinivalue_line("filterwarnings", warning_line)
+
+    set_log_level_mne("WARNING")
+    set_log_level("WARNING")
```
Comment (mscheltienne) [inline pycrostates/conftest.py:28]:
```suggestion

    logger.propagate = True
    set_log_level_mne("WARNING")
    set_log_level("WARNING")
```

**Comment by vferat** [comment]
Hey @mscheltienne,

I've activated the **pre_commit** for this repository.

Regarding your suggestion to drop the Azure Pipelines (AZP) workflows in favor of GitHub Actions (GH), I believe it's a solid idea. It can make it easier for contributors to manage our automation, and Ialways happy to reduce reliance on external third-party services.
We need to keep in mind that we'll be limited to using only one **github runner** ( except if we find a sponsorship). Let's give it  a shot

Thanks for your work in this PR, everything else looks good to me !

**Comment by vferat** [comment]
`pre-commit.ci` seems to be running smoothly

File: pycrostates/conftest.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    config.addinivalue_line("usefixtures", "matplotlib_config")
+
+    warnings_lines = r"""
+    error::
+    # We use matplotlib agg backend to avoid any window to pop up during tests.
+    ignore:Matplotlib is currently using agg:UserWarning
+    # Pytest internals
+    ignore:Use setlocale.*instead:DeprecationWarning
+    ignore:datetime\.datetime\.utcnow.*is deprecated.*:DeprecationWarning
+    """
```
Comment (mscheltienne) [inline pycrostates/conftest.py:22]:
```suggestion
    ignore:Use setlocale.*instead:DeprecationWarning
    ignore:datetime\.datetime\.utcnow.*is deprecated.*:DeprecationWarning
    ignore:datetime\.datetime\.utcfromtimestamp.*is deprecated.*:DeprecationWarning
    """
```

**Comment by mscheltienne** [comment]
> we'll be limited to using only one github runner ( except if we find a sponsorship)

What do you mean? GitHub self-hosted runners are unlimited for public repositories.
c.f. https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions

> GitHub Actions usage is free for standard GitHub-hosted runners in public repositories, and for self-hosted runners.

---------

That should be the last warning to fix, from `joblib` https://github.com/joblib/joblib/pull/1518
I'll merge once green, and we can always revert to Azure if we want to later.

**Pull request merged**